warp-core-macros/src/main/scala/com/workday/warp/monadic/WarpMacros.scala

```diff
@@ -0,0 +1,149 @@
+package com.workday.warp.monadic
+
+import com.workday.warp.monadic.WarpAlgebra.WarpScript
+
+import language.experimental.macros
+import scala.reflect.macros.blackbox
+
+/**
+ * Created by tomas.mccandless on 8/16/21.
+ */
+object WarpMacros {
+
+  /**
+    * A simple toy macro that adds 2 integers
+    *
+    * @param a
+    * @param b
+    * @return the sum of `a` and `b`.
+    */
+  def add(a: Int, b: Int): Int = macro addImpl
+
+  def addImpl(ctx: blackbox.Context)(a: ctx.Expr[Int], b: ctx.Expr[Int]): ctx.Expr[Int] = {
+    import ctx.universe.reify
+    // reify lifts a value into scala AST, Int => Expr[Int]
+    // splice (in the context of reify) converts from scala AST -> value, ie Expr[Int] => Int
+    reify {
```

Comment:
good question, its a bit confusing. in a macro implementation, we can't do just `a + b`, because they aren't `Int`, they are `Expr[Int]` (so they have no `+` method). 

Here is some documentation on `reify` and `splice`:

```
Use reify to produce the abstract syntax tree representing a given Scala expression.
For example:

 val five = reify{ 5 }         // Literal(Constant(5))
 reify{ 5.toString }           // Apply(Select(Literal(Constant(5)), TermName("toString")), List())
```


```
Splice is a dummy method to mark expression splicing in reification.
It should only be used within a reify call, which eliminates the splice call and embeds the wrapped tree into the reified surrounding expression. If used alone splice throws an exception when called at runtime.
If you want to use an Expr in reification of some Scala code, you need to splice it in. For an expr of type Expr[T], where T has a method foo, the following code:

   reify{ expr.splice.foo }
 
uses splice to turn an expr of type Expr[T] into a value of type T in the context of reify.
```

so basically at compile-time, we are rewriting an AST representing `add(a, b)` into an AST representing `a + b`. hope that clears it up, i just left this add macro here as a simpler example

Reply:
Just to add on, note that `reify` is essentially what slick is doing when lifting from a primitive type, i.e `Int` to `Rep[Int]`